src/textual/devtools.py

```diff
@@ -0,0 +1,265 @@
+            info_renderable = DevtoolsInternalMessage(
+                f"Discarded {spillover} messages", level="warning"
+            )
+            console.print(info_renderable)
+        incoming_queue.task_done()
+
+
+async def _consume_outgoing(
+    outgoing_queue: Queue[dict], websocket: WebSocketResponse
+) -> None:
+    while True:
+        message_json = await outgoing_queue.get()
+        type = message_json["type"]
+        if type == "server_info":
+            await websocket.send_json(message_json)
+
+
+async def websocket_handler(request: Request):
+    """aiohttp websocket handler for sending data between devtools client and server"""
+
+    websocket = WebSocketResponse()
+    await websocket.prepare(request)
+    request.app["websockets"].add(websocket)
+
+    console = request.app["console"]
+    size_change_poll_delay = request.app["size_change_poll_delay_secs"]
+
+    incoming_queue: Queue[dict] = Queue()
+    outgoing_queue: Queue[dict] = Queue()
+
+    request.app["tasks"].extend(
+        (
+            asyncio.create_task(_consume_outgoing(outgoing_queue, websocket)),
+            asyncio.create_task(
+                _enqueue_size_changes(
+                    console, outgoing_queue, poll_delay=size_change_poll_delay
+                )
+            ),
+            asyncio.create_task(_consume_incoming(console, incoming_queue)),
+        )
+    )
+
+    console.print(
+        DevtoolsInternalMessage(f"Client '{escape(request.remote)}' connected")
+    )
+
+    await _enqueue_server_info(
+        outgoing_queue, width=console.width, height=console.height
+    )
+    try:
+        async for message in websocket:
+            message = cast(WSMessage, message)
+            if message.type == WSMsgType.TEXT:
+                try:
+                    message_json = json.loads(message.data)
+                except JSONDecodeError:
+                    console.print(escape(str(message.data)))
+                    continue
+
+                type = message_json.get("type")
+                if not type:
+                    continue
+                if type in QUEUEABLE_TYPES:
+                    await incoming_queue.put(message_json)
+            elif message.type == WSMsgType.ERROR:
+                console.print(
+                    DevtoolsInternalMessage("Websocket error occurred", level="error")
+                )
+                break
+    except Exception as error:
+        console.print(DevtoolsInternalMessage(str(error), level="error"))
+    finally:
+        request.app["websockets"].discard(websocket)
+        console.print()
+        console.print(
+            DevtoolsInternalMessage(f"Client '{escape(request.remote)}' disconnected")
+        )
+
+    return websocket
+
+
+async def _on_shutdown(app: Application) -> None:
+    for task in app["tasks"]:
+        task.cancel()
+        with suppress(CancelledError):
+            await task
+
+    for websocket in set(app["websockets"]):
```

Comment:
Haven't you already killed the task that was processing the websockets by this point? I think this would mean that the websocket isn't closed gracefully.

Reply:
I've already cancelled the task yes, but not the socket. I don't believe task cancellation would close the socket though, so this ensures that after the tasks are cancelled the websockets are also closed.